gateway/gateway.go

```diff
@@ -17,34 +17,111 @@ type Config struct {
+// From >= 0 and To = nil: Get the file (From, Length)
+// From >= 0 and To >= 0: Get the range (From, To)
+// From >= 0 and To <0: Get the range (From, Length - To)
+type GetRange struct {
+	From uint64
+	To   *int64
+}
+
+// API is the required set of functionality used to implement the IPFS HTTP Gateway specification.
+// To signal error types to the gateway code (so that not everything is a 500) return a (wrapped) error defined in this package.
+// There are also some existing error types that the gateway code knows how to handle (e.g. context.DeadlineExceeded
+// and various IPLD pathing related errors).
 type API interface {
-	// GetUnixFsNode returns a read-only handle to a file tree referenced by a path.
-	GetUnixFsNode(context.Context, path.Resolved) (files.Node, error)
+	// Get returns a UnixFS file, UnixFS directory, or an IPLD block depending on what the path is that has been
+	// requested. Directories' files.DirEntry objects do not need to contain content, but must contain Name,
+	// Size, and Cid.
+	// TODO: Should we export or just better describe the Cid metadata interface implemented in go-unixfs?
+	Get(context.Context, ImmutablePath) (ContentPathMetadata, files.Node, error)
 
-	// LsUnixFsDir returns the list of links in a directory.
-	LsUnixFsDir(context.Context, path.Resolved) (<-chan iface.DirEntry, error)
+	// GetRange returns a full UnixFS file object, however reading from values outside the given ranges can error
+	GetRange(context.Context, ImmutablePath, ...GetRange) (ContentPathMetadata, files.File, error)
 
-	// GetBlock return a block from a certain CID.
-	GetBlock(context.Context, cid.Cid) (blocks.Block, error)
+	// GetAll returns a UnixFS file or directory depending on what the path is that has been requested. Directories should
+	// include all content recursively.
+	GetAll(context.Context, ImmutablePath) (ContentPathMetadata, files.Node, error)
+
+	// GetBlock returns a single block of data
+	GetBlock(context.Context, ImmutablePath) (ContentPathMetadata, files.File, error)
+
+	// Head returns a file or directory depending on what the path is that has been requested.
+	// For UnixFS files should return a file which has the correct file size and either returns the content type or
+	// enough data (e.g. 3kiB) such that the content type can be determined by sniffing.
+	// For all other data types returning just size information is sufficient
+	// TODO: give function more explicit return types
+	Head(context.Context, ImmutablePath) (ContentPathMetadata, files.Node, error)
+
+	// ResolvePath resolves the path using UnixFS resolver. If the path does not
+	// exist due to a missing link, it should return an error of type:
+	// https://pkg.go.dev/github.com/ipfs/go-path@v0.3.0/resolver#ErrNoLink
+	// TODO: Should we be more liberal in IPLD pathing errors returned here and/or provide some tooling to help unify errors here?
+	ResolvePath(context.Context, ImmutablePath) (ContentPathMetadata, error)
```

Comment:
We don't want to depend on external error types, maybe something like:

```suggestion
	// ResolvePath resolves the path using UnixFS resolver. If the path does not
	// exist due to a missing link, it should return an error of type:
	// NewErrorResponse(fmt.Errorf("no link named %q under %s", name, cid), http.StatusNotFound)
	ResolvePath(context.Context, ImmutablePath) (ContentPathMetadata, error)
```